# Promise

## 基础用法

### Promise.prototype.then

`then`方法返回的是一个**新的**`Promise`实例。

### Promise.prototype.catch

catch 方法是`.then(null, rejection)`、`.then(undefined, rejection)`的语法糖，用于发生错误时的回调。

通常不会使用 then 的第二个参数来捕获错误，理由是 catch 方法可以捕获之前任意 then 方法中的错误，写法更接近 try/catch 的用法。

### Promise.resolve

该方法可以将一个对象转为 Promise 对象。

```js
Promise.resolve('foo')
// 等价于
new Promise(resolve => resolve('foo'))
```

1）参数是一个 Promise对象，则直接返回该实例

2）如果参数是 thenable 对象（具有 then 方法的对象），则会将该对象转换为 Promise 对象并立即执行 then 方法。返回一个状态为 resolved 的新 Promise 对象。

```js
let thenable = {
  then: function(resolve, reject) {
    resolve(42);
  }
};

let p1 = Promise.resolve(thenable);
p1.then(function(value) {
  console.log(value);  // 42
});
```

3）参数不是对象或不是 thenable 对象，则会返回一个状态为 resolved 的新 Promise 对象。

```js
const p = Promise.resolve('Hello');

p.then(function (s){
  console.log(s)
});
// Hello
```

4）无参数，直接返回一个 resolve 状态的 Promise 对象。且该 resolve() 是在本轮 event loop 结束时执行。

### Promise.reject

该方法也会返回一个新的 Promise 实例，状态为 rejected。

```js
const p = Promise.reject('出错了');
// 等同于
const p = new Promise((resolve, reject) => reject('出错了'))

p.then(null, function (s) {
  console.log(s)
});
// 出错了
```

与 Promise.resolve() 不同的是，该方法的参数会直接**作为 reject 的返回值**，不作任何处理。

### Promise.prototype.finally

用于指定不论 Promise 对象最后状态如何都会执行的操作。返回**原来的值**。

```js
promise.then(res => {}).finally(() => {})
```

该方法不接受任何参数，即无法得知前面的 Promise 状态。因此，finally 中的操作应该是与**状态无关**的，**不依赖**于 Promise 的执行结果。

```js
Promise.resolve(2).finally(() => {}).then(res => { console.log(res) }) // 2
```

finally 本质上是 then 方法的特例。

```js
promise.finally(() => {
	// do sth
})
// 等同于
promise.then(res => {
  // do sth
  return res
}, error => {
  // do sth
  throw error
})
```

### Promise.all

用于多个 Promise 实例，包装成一个新的 Promise 实例。方法接收一个数组作为参数，每一项都要是 Promise 实例，如果不是则调用 Promise.resolve 进行转换。入参可以不是数组，但需要有 Iterator 接口。

只有当数组中的每一项都变为 fulfilled 时，最终的状态才会变成 fulfilled，返回值仍为数组。当其中一项被 rejected，则最终的状态变为 rejected，此时返回**第一个被 reject 的实例**返回值。

如果作为参数的 promise 实例有自己定义的catch方法，则不会触发promise.all的catch方法。最终导致返回的实例中会包含catch返回的promise。如果参数都没有自己的catch方法，才会调用promise.all的catch方法。

如果想要捕获参数实例中的错误并在返回数组中体现：

```js
Promise.all([request1.catch(err => err),  request2.catch(err => err)])
	.then((res) => {
  		console.log(res)
	})
```

参考 Promise A+ 规范，只要 onFulfilled 或者 onRejected 返回一个值 x ， 后一个 promise 都会进入 onFulfilled 状态。此方法可以来捕获请求错误的处理。

### Promise.race

```js
const p = Promise.race([p1, p2, p3]);
```

该方法对于参数的处理与all方法一样。同样接受一个promise数组，只要其中一个参数实例改变状态，则该那个改变的promise实例返回值会最为p的回调。

### Promise.allSettled

该方法接收一组 Promise 实例作为参数，包装为一个新的 Promise 实例。只有当这些参数实例都返回结果，无论fulfilled还是rejected，包装实例才会结束。一旦结束，状态总是fulfilled。

```js
const resolved = Promise.resolve(42);
const rejected = Promise.reject(-1);

const allSettledPromise = Promise.allSettled([resolved, rejected]);

allSettledPromise.then(function (results) {
  console.log(results);
});
// [
//    { status: 'fulfilled', value: 42 },
//    { status: 'rejected', reason: -1 }
// ]
```

每个实例的结果都含有status属性，该属性的值只可能是字符串fulfilled或者rejected。

该方法用于不关心异步操作的结果，**只关心这些操作是否结束**。相比于Promise.all方法，后者无法做到这一点（一个参数实例rejected，结果就会变成rejected）。

### Promise.any

接收一组 Promise 实例作为参数。只要其中一个实例变成 fulfilled 状态，最终结果就会变成fulfilled。如果所有实例都变成rejected状态，则最终结果也会变成rejected状态。

## Promise A+ 规范

### 三种状态

promise 可能有三种状态：等待（pending）、已完成（fulfilled）、已拒绝（rejected）。

promise 的状态只可能从“等待”转为另外两者，不能够逆向转换。且一旦改变后状态一直保持这个结果不会再发生改变。

### then 方法

**（重要）**then 方法必须返回一个 promise 对象 `promise2 = promise1.then(onFulfilled, onRejected); `

+ 只要 onFulfilled 或者 onRejected 返回一个值 x ，promise 2 都会进入 onFulfilled 状态
+ 如果 onFulfilled 或者 onRejected 抛出一个异常 e ，则 promise2 必须拒绝执行，并返回 拒因 e
+ 如果 onFulfilled 不是函数且 promise1 状态变为已完成， promise2 必须成功执行并返回相 同的值
+ 如果 onRejected 不是函数且 promise1 状态变为已拒绝， promise2 必须执行拒绝回调并返 回相同的据因 

```js
var promise1 = new Promise((resolve, reject) => {reject();}); 
promise1
	.catch(err => {
  	throw Error('123')
  	// console.log('123')
  	// return 123
  	// 上述两种情况都会走到下一个 then 的第一个参数回调中 （第一点）
	})
	.then(null, function() { // 进入第二个回调函数参数
  	console.log('errr')
  	return 123;
	}) // onfulfilled (参考第一点)
	.then(null, null) // 参考第三点
  .then(null, null) // 参考第三点
  .then(
		() => {
		console.log('promise2 已完成');
	},
	() => {
	console.log('promise2 已拒绝'); }
);
```



## 实现源码

### finally

该方法会返回原来的值。

```js
Promise.finally = function (callback) {
  let p = this.constructor
  return this.then(
  	res => P.resolve(callback()).then(() => res),
    err => P.resolve(callback()).then(() => err)
  )
}
```

### all

```js
Promise.myAll = function(promises) {
  let result = [];
  let count = 0;
  return new Promise((resolve, reject) => {
    for (let i = 0; i < promises.length; i++) { // 最好使用 forEach 来控制并发
      const curPromise = promises[i];
      curPromise.then(res => {
        result[i] = res;
        count++;
        if (count === promises.length) {
          resolve(result);
        }
      }, reject);
    }
  });
};
```

### race

```js
Promise.myRace = function(promises) {
  return new Promise((resolve, reject) => {
    for (let i = 0; i < promises.length; i++) { // 最好使用 forEach 来控制并发
      const curPromise = promises[i];
      curPromise.then(resolve, reject);
    }
  });
};
```

## 十道题

+ [练习题](https://juejin.im/post/5a04066351882517c416715d)

# asycn/await

